server/internal/core/application/indexer.go

```diff
@@ -0,0 +1,553 @@
+				totalBatchAmount += receiver.Amount
+			}
+		}
+	}
+
+	batches := make(map[VOut]Batch)
+	// TODO: currently commitment tx has only one batch, in future multiple batches will be supported
+	batches[0] = Batch{
+		TotalBatchAmount:   totalBatchAmount,
+		TotalForfeitAmount: totalForfeitAmount,
+		TotalInputVtxos:    totalInputVtxos,
+		TotalOutputVtxos:   int32(len(leaves)),
+		ExpiresAt:          vtxo[0].ExpireAt,
+		Swept:              round.Swept,
+	}
+
+	return &CommitmentTxResp{
+		StartedAt: round.StartingTimestamp,
+		EndAt:     round.EndingTimestamp,
+		Batches:   batches,
+	}, nil
+}
+
+func (i *indexerService) GetVtxoTree(ctx context.Context, req VtxoTreeReq) (*VtxoTreeResp, error) {
+	vtxoTree, err := i.repoManager.Rounds().GetVtxoTreeWithTxid(ctx, req.BatchOutpoint.Txid) //TODO repo methods needs to be updated with multiple batches in future
+	if err != nil {
+		return nil, err
+	}
+
+	nodes, pageResp := paginate(flattenNodes(vtxoTree), req.Page, maxPageSizeVtxoTree)
+
+	return &VtxoTreeResp{
+		Nodes: nodes,
+		Page:  pageResp,
+	}, nil
+}
+
+func (i *indexerService) GetForfeitTxs(ctx context.Context, req ForfeitTxsReq) (*ForfeitTxsResp, error) {
+	round, err := i.repoManager.Rounds().GetRoundWithTxid(ctx, req.BatchOutpoint.Txid) //TODO batch thing
+	if err != nil {
+		return nil, err
+	}
+
+	txs := make([]string, 0, len(round.ForfeitTxs))
+	for _, tx := range round.ForfeitTxs {
+		txs = append(txs, tx.Txid)
+	}
+
+	forfeitTxs, pageResp := paginate(txs, req.Page, maxPageSizeForfeitTxs)
+
+	return &ForfeitTxsResp{
+		Txs:  forfeitTxs,
+		Page: pageResp,
+	}, nil
+
+}
+
+func (i *indexerService) GetConnectors(ctx context.Context, req ConnectorsReq) (*ConnectorResp, error) {
+	round, err := i.repoManager.Rounds().GetRoundWithTxid(ctx, req.BatchOutpoint.Txid) //TODO batch thing
+	if err != nil {
+		return nil, err
+	}
+
+	connectors, pageResp := paginate(flattenNodes(round.Connectors), req.Page, maxPageSizeConnector)
+
+	return &ConnectorResp{
+		Connectors: connectors,
+		Page:       pageResp,
+	}, nil
+}
+
+func (i *indexerService) GetSpendableVtxos(ctx context.Context, req SpendableVtxosReq) (*SpendableVtxosResp, error) {
+	decodedAddress, err := common.DecodeAddress(req.Address)
+	if err != nil {
+		return nil, err
+	}
+
+	if !bytes.Equal(schnorr.SerializePubKey(decodedAddress.Server), schnorr.SerializePubKey(i.pubkey)) {
+		return nil, err
+	}
+
+	pubkey := hex.EncodeToString(schnorr.SerializePubKey(decodedAddress.VtxoTapKey))
+
+	spendableVtxos, _, err := i.repoManager.Vtxos().GetAllNonRedeemedVtxos(ctx, pubkey)
+	if err != nil {
+		return nil, err
+	}
+
+	spendableVtxosPaged, pageResp := paginate(spendableVtxos, req.Page, maxPageSizeSpendableVtxos)
+
+	return &SpendableVtxosResp{
+		Vtxos: spendableVtxosPaged,
+		Page:  pageResp,
+	}, nil
+}
+
+func (i *indexerService) GetTransactionHistory(ctx context.Context, req TxHistoryReq) (*TxHistoryResp, error) {
+	allVtxos, err := i.repoManager.Vtxos().GetAll(ctx)
```

Comment:
let's make use of an ad hoc API `GetVtxosWithPubKey` instead of getting the whole vtxo set and then filter by pubkey. The code to extract the pubkey from the address can be taken from the func above

Reply:
Done